twister: increase resilience when running outside of git.

Flameeyes opened the pull request:
twister: increase resilience when running outside of git.



While the code already includes some provisions to allow running outside
of a git checkout, attempting that right now causes a failure to generate
the XML report, as the JSON report creates a null `zephyr_version` field.

The reason for that is that the original code doesn't set `self.version`
when the subprocess returns a non-zero status (and also doesn't log.)

Instead of having to set `self.version` for all failure cases, initialize
it to the failure state, and log if it hasn't changed from that.

Signed-off-by: Diego Elio Pettenò <flameeyes@meta.com>

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/zephyrproject-rtos/zephyr/pull/61668).
* #61728
* __->__ #61668
* #61736

ycsin reviewed (dismissed):
I also observed this issue in my non-git build and verified that this fixed the issue

PerMac commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -834,20 +834,18 @@ def check_zephyr_version(self):
-                    self.version = "Unknown"
-                    logger.error("Could not determine version")
         except OSError:
-            logger.info("Cannot read zephyr version.")
```
From what I know, `Except: pass` constructs are to be avoided, since they can hide a real issue. I believe a logging should stay, at least at a debug level. Maybe like this?
```
except OSError as e:
    logger.info(f"Cannot read zephyr version: {str(e)}")
```

Flameeyes replied:
To be honest, the likeliness of an `OSError` on `subprocess.run` is so low that you'd probably never be hitting it in the first place, and I was of two minds of not catching it at all, and just let it bubble up — it is not caught on the following call for `git show` so I'm not quite sure why this is treated differently.

Though we should really use `logging.exception("Failure when reading Zephyr version")` to print the stack trace if we want the log.

Which one do you prefer?

PerMac replied:
I opt for `logger.exception`. Thanks

PerMac reviewed (dismissed):
Thanks!

Flameeyes committed:
twister: increase resilience when running outside of git.



While the code already includes some provisions to allow running outside
of a git checkout, attempting that right now causes a failure to generate
the XML report, as the JSON report creates a null `zephyr_version` field.

The reason for that is that the original code doesn't set `self.version`
when the subprocess returns a non-zero status (and also doesn't log.)

Instead of having to set `self.version` for all failure cases, initialize
it to the failure state, and log if it hasn't changed from that.

Signed-off-by: Diego Elio Pettenò <flameeyes@meta.com>

Flameeyes committed:
uart_sam0: fix shadowed variable.

This code was added to CI after the `-Wshadow` stack was prepared, so it
was not fixed together with the rest.

Signed-off-by: Diego Elio Pettenò <flameeyes@meta.com>

nashif commented:
should have sent the shadow fix in a seperate PR....

Flameeyes commented:
> should have sent the shadow fix in a separate PR....

I did, #61736

But because I needed that fix to make sure this was still working I also stacked them together — it gets elided when rebase merging anyway.
